include/xplugin/xshared_library.hpp

```diff
@@ -1,55 +1,162 @@
 /***************************************************************************
 * Copyright (c) 2018, Johan Mabille, Sylvain Corlay, Wolf Vollprecht and   *
-* Martin Renou                                                             *
+* Martin Renou, Dr. Thorsten Beier                                         *
 *                                                                          *
 * Distributed under the terms of the BSD 3-Clause License.                 *
 *                                                                          *
 * The full license is in the file LICENSE, distributed with this software. *
 ****************************************************************************/
-
+#pragma once
 #include <string>
 
 #include "xplugin_config.hpp"
 
+#if (defined(__linux__) || defined(__unix__) || defined(__APPLE__))
+#include <dlfcn.h>
+#include "xplugin/xshared_library.hpp"
+#endif
+
+#ifdef _WIN32
+#include <windows.h>
+#include "xplugin/xshared_library.hpp"
+#endif
+
+
+
 namespace xp
 {
-    class XPLUGIN_API xshared_library
+
+
+    #if (defined(__linux__) || defined(__unix__) || defined(__APPLE__))
+    class xunix_shared_library 
     {
     public:
+        // delete copy constructor
+        xunix_shared_library(const xunix_shared_library&) = delete;
+        // delete copy assignment
+        xunix_shared_library& operator=(const xunix_shared_library&) = delete;
 
-        virtual ~xshared_library() {}
-
-        xshared_library(const xshared_library&) = delete;
-        xshared_library& operator=(const xshared_library&) = delete;
+        // move constructor
+        xunix_shared_library(xunix_shared_library&& other) noexcept = default;
+        // move assignment
+        xunix_shared_library& operator=(xunix_shared_library&& other) noexcept = default;
 
-        xshared_library(xshared_library&&) = delete;
-        xshared_library& operator=(xshared_library&&) = delete;
 
-        static xshared_library* open(const std::string& name);
-        static void close(xshared_library* library);
-        static const char* get_error();
+        xunix_shared_library(std::filesystem::path path)
+        :   m_path(path),
+            m_handle(nullptr)
+        {
+           this->open();
+        }
```

Comment:
but we **not want to close** in the destructor. This result in segfaults when using already created factories/instances.
Except if we use the `RTLD_NODELETE` flag in `dlopen`. But then then, the library cannot be closed at all.

Reply:
If we want to close in the destructor, the `xplugin_registry` would have to be used as a singleton

Reply:
Why that? the destructor would be called only if you remove the library from the plugin registry, right? If someone decides to add the same library to many registries, it is his responsibility; as it would be if he decided to call twice dlopen or dlclose directly.